File: lib/iriq/cluster.rb
module Iriq
  # A group of identifiers that share a host + shape key. Tracks examples and
  # per-position segment statistics so callers can ask which positions are
  # actually stable in practice (e.g. /users/ always literal, /{integer}
  # always variable).
  class Cluster
    attr_reader :key, :host, :scheme, :shape, :examples, :count, :param_stats, :max_values

    # Structured Shape lazily derived from the first observed example —
    # Iriq::Shape, or nil if no examples are present yet. Cached after the
    # first call.
    def shape_object(classifier: SegmentClassifier::DEFAULT)
      return @shape_object if @shape_object
      return nil if @examples.empty?

      @shape_object = Shape.from_segments(@examples.first.path_segments, classifier: classifier)
    end

    MAX_EXAMPLES = 10

    # Share of date-typed observations required before the corpus promotes
    # a param to :date. 8-digit IDs in the 1900..2100 range look like
    # YYYYMMDD by accident — without quorum we'd canonicalize random IDs.
    DATE_CONFIDENCE_THRESHOLD = 0.8

    # `:number` umbrella thresholds. Promote a position to :number when
    # the combined :integer + :float observations dominate (≥ majority)
    # AND neither subtype alone hits the strong threshold (we have a clear
    # numeric pattern but it isn't purely ints or purely floats).
    NUMBER_CONFIDENCE_THRESHOLD = 0.8
    NUMBER_SUBTYPE_THRESHOLD    = 0.8

    # Param classification is a confidence ladder: constant → string → enum.
    # A param with a single observed value is a constant (rendered as-is); one
    # that varies but isn't yet a trustworthy enum is :string (a generic
    # placeholder); a bounded, well-supported value set is :enum.
    #
    # `:enum` thresholds. Promote a param to :enum when the corpus has seen
    # enough samples to trust the bound (ENUM_MIN_OBSERVATIONS), the *established*
    # values — those seen at least ENUM_MIN_VALUE_COUNT times — are few
    # (ENUM_MAX_CARDINALITY) and cover nearly all observations
    # (ENUM_MIN_COVERAGE). Rare one-off values are stragglers, not
    # disqualifiers: this is what keeps a single brand-new value from knocking
    # an established enum back down (the observe-before-normalize case).
    ENUM_MIN_OBSERVATIONS = 20
    ENUM_MAX_CARDINALITY  = 10
    ENUM_MIN_VALUE_COUNT  = 3
    ENUM_MIN_COVERAGE     = 0.9
    # An enum is a bounded *set*: a single repeated value is a constant, not an
    # enum, so it takes at least two established members to qualify.
    ENUM_MIN_MEMBERS      = 2

    # `:string` — a param that has taken on 2+ distinct non-typed values but
    # isn't (yet) a confident enum. The intermediate rung: we know it varies
    # and looks like free-form text, but haven't earned the bounded-set claim.
    STRING_MIN_DISTINCT = 2

    # Confidence smoothing constant. confidence = total / (total + K): a
    # monotone curve that is 0.5 at K observations and asymptotes to 1.0. The
    # type names our guess; this number says how much evidence backs it.
    CONFIDENCE_SMOOTHING = 15

    def initialize(key:, host:, scheme:, shape:, max_values: PositionStats::DEFAULT_MAX_VALUES)
      @key            = key
      @host           = host
      @scheme         = scheme
      @shape          = shape
      @shape_object   = nil
      @examples       = []
      @example_keys   = Set.new
      @count          = 0
      @segment_counts = []
      @max_values     = max_values
      # Query-param stats keyed by param name. Each is a PositionStats — same
      # cardinality cap, same type-counts machinery, just indexed by ?key=
      # instead of by path position.
      @param_stats    = {}
    end

    def add(identifier, classifier: SegmentClassifier::DEFAULT)
      @count += 1
      if @examples.size < MAX_EXAMPLES
        canon = identifier.canonical
        @examples << identifier unless @example_keys.include?(canon)
        @example_keys << canon
      end

      identifier.path_segments.each_with_index do |seg, i|
        @segment_counts[i] ||= Hash.new(0)
        @segment_counts[i][seg] += 1
      end

      return unless identifier.query_params
      identifier.query_params.each do |name, value|
        stats = @param_stats[name] ||= PositionStats.new(max_values: @max_values)
        value_s = value.to_s
        stats.observe(value_s, classifier.classify(value_s))
      end
    end

    # Per-position summary:
    #   [
    #     { position: 0, stable: true,  values: { "users" => 3 } },
    #     { position: 1, stable: false, values: { "1" => 1, "2" => 1, "3" => 1 } },
    #   ]
    def segment_stats
      @segment_counts.each_with_index.map do |counts, i|
        {
          position: i,
          stable:   counts.size == 1,
          values:   counts.dup,
        }
      end
    end

    def to_h
      {
        key:      key,
        host:     host,
        scheme:   scheme,
        shape:    shape,
        count:    count,
        examples: examples.map(&:canonical),
        segments: segment_stats,
        params:   param_summary,
      }
    end

    # Per-param summary, ordered by descending presence. Each entry is:
    #   { name: "page", count: N, type: :integer, confidence: 0.83,
    #     cardinality: K, presence: 0.83 }
    # confidence is how much evidence backs the type (see param_confidence);
    # presence is count / @count — the fraction of observations that had
    # this param.
    def param_summary
      return [] if @param_stats.empty?

      @param_stats.map { |name, _stats|
        stats = @param_stats[name]
        type  = param_type(name)
        row   = {
          name:        name,
          count:       stats.total,
          type:        type,
          confidence:  param_confidence(stats),
          cardinality: stats.cardinality,
          presence:    @count.positive? ? stats.total.to_f / @count : 0.0,
        }
        row[:values] = enum_values(stats) if type == :enum
        # Verbose value distribution — fractions over tracked occurrences.
        # Boolean and enum positions get the per-value breakdown (e.g.
        # `true: 0.97, false: 0.03`). Number positions get the int-vs-float
        # split via :subtype_distribution.
        if type == :boolean || type == :enum
          row[:value_distribution] = value_distribution(stats)
        end
        if type == :number
          row[:subtype_distribution] = subtype_distribution(stats, %i[integer float])
        end
        # :file kind breakdown — derived from tracked value_counts at
        # summary time. Best-effort: only reflects observations within
        # the value-tracking cap.
        if type == :file
          row[:kind_distribution] = file_kind_distribution(stats)
        end
        if stats.numeric_count.positive?
          row[:min] = stats.numeric_min
          row[:max] = stats.numeric_max
          row[:avg] = stats.numeric_avg
        end
        row
      }.sort_by { |row| [-row[:count], row[:name]] }
    end

    # Returns the type the corpus is confident enough to call this param.
    # Equals stats.dominant_type when the dominant type isn't :date; when
    # :date is dominant but below DATE_CONFIDENCE_THRESHOLD, falls back to
    # the most-common non-date type (or :literal if none exists). Shared
    # by Cluster#param_summary and Corpus#inferred_param_type so both views
    # agree on what the corpus "thinks" about a param.
    def param_type(name)
      stats = @param_stats[name]
      return nil unless stats
      return nil if stats.total.zero?

      type = stats.dominant_type

      # :year takes priority over :enum for numeric range columns —
      # a "years 2020..2026" position is more useful described as a
      # ranged year than as an enum of those specific values.
      return :year if year_position?(type, stats)
      # :http_status — 3-digit ints clustered in 100..599 are almost
      # certainly HTTP statuses. Same shape as :year (range check) but
      # tighter window. Useful for `?status=...` or path positions that
      # echo a status code.
      return :http_status if http_status_position?(type, stats)

      # :enum check — bounded set of repeated values trumps the underlying
      # value type. `?status=active|draft|archived` surfaces as :enum
      # (with the value list) rather than :literal even though each value
      # individually classifies as a literal. Skip the override when the
      # dominant type is already specific (`:boolean` carries more meaning
      # than a 2-value enum).
      return :enum if enum?(stats) && type != :boolean

      # :date gate — demote when there isn't enough date-typed quorum.
      if type == :date
        date_frac = stats.type_counts[:date].to_f / stats.total
        return type if date_frac >= DATE_CONFIDENCE_THRESHOLD

        return dominant_excluding(stats, :date) || :literal
      end

      # :number umbrella — promote when ints + floats together dominate
      # but neither alone is the clear winner.
      if type == :integer || type == :float
        int_frac   = stats.type_counts[:integer].to_f / stats.total
        float_frac = stats.type_counts[:float].to_f / stats.total
        if int_frac < NUMBER_SUBTYPE_THRESHOLD &&
           float_frac < NUMBER_SUBTYPE_THRESHOLD &&
           (int_frac + float_frac) >= NUMBER_CONFIDENCE_THRESHOLD
          return :number
        end
      end

      # Param-name fallback — `?phone=...` overrides a generic literal
      # type with `:phone` when the value's shape was too weak to detect
      # on its own. Only fires for overridable types (literal/opaque_id/slug).
      if (hint = SegmentClassifier.param_name_hint(name, type))
        return hint
      end

      # :string rung — a literal-valued param that has taken on more than one
      # distinct value varies, so it's a placeholder, not a fixed constant.
      # Below the enum bar (checked above), so we claim only "free-form text".
      return :string if type == :literal && stats.cardinality >= STRING_MIN_DISTINCT

      type
    end

    YEAR_RANGE              = 1900..2100
    YEAR_MIN_OBSERVATIONS   = 5
    YEAR_MIN_DISTINCT       = 2
    YEAR_MAX_DISTINCT       = 150

    def year_position?(type, stats)
      return false unless type == :integer
      return false if stats.numeric_count.zero?
      return false if stats.cardinality < YEAR_MIN_DISTINCT
      return false if stats.cardinality > YEAR_MAX_DISTINCT
      return false if stats.total < YEAR_MIN_OBSERVATIONS

      YEAR_RANGE.cover?(stats.numeric_min) && YEAR_RANGE.cover?(stats.numeric_max)
    end

    HTTP_STATUS_RANGE            = 100..599
    HTTP_STATUS_MIN_OBSERVATIONS = 5
    HTTP_STATUS_MIN_DISTINCT     = 2
    HTTP_STATUS_MAX_DISTINCT     = 30

    def http_status_position?(type, stats)
      return false unless type == :integer
      return false if stats.numeric_count.zero?
      return false if stats.cardinality < HTTP_STATUS_MIN_DISTINCT
      return false if stats.cardinality > HTTP_STATUS_MAX_DISTINCT
      return false if stats.total < HTTP_STATUS_MIN_OBSERVATIONS

      HTTP_STATUS_RANGE.cover?(stats.numeric_min) && HTTP_STATUS_RANGE.cover?(stats.numeric_max)
    end

    # True when stats shows a bounded set of repeated values worth treating
    # as an enum. Built around the *established* members (values seen at least
    # ENUM_MIN_VALUE_COUNT times) so a stray one-off value is a straggler, not
    # a disqualifier. See ENUM_* constants at the top of this class.
    def enum?(stats)
      return false if stats.total < ENUM_MIN_OBSERVATIONS

      established = established_values(stats)
      return false unless established.size.between?(ENUM_MIN_MEMBERS, ENUM_MAX_CARDINALITY)

      coverage = established.values.sum.to_f / stats.total
      coverage >= ENUM_MIN_COVERAGE
    end

    # Values seen often enough to count as real members of the set (vs noise).
    def established_values(stats)
      stats.value_counts.select { |_, n| n >= ENUM_MIN_VALUE_COUNT }
    end

    # Confidence that the assigned type is right, given how much evidence backs
    # it. Monotone in observation count; 0.5 at CONFIDENCE_SMOOTHING, → 1.0.
    def param_confidence(stats)
      return 0.0 if stats.total.zero?

      (stats.total.to_f / (stats.total + CONFIDENCE_SMOOTHING)).round(2)
    end

    # The enum's member values — the established ones (seen enough to be real),
    # ordered by descending count (lex tie-break). Stragglers are excluded so
    # the advertised set is what the corpus is actually confident about.
    def enum_values(stats)
      established_values(stats).sort_by { |v, n| [-n, v] }.map(&:first)
    end

    # value_distribution returns the fraction of total observations each
    # tracked value represents, ordered by descending count then lex. Used
    # by param_summary for :boolean and :enum positions so callers can
    # render "true 97%, false 3%"-style breakdowns.
    def value_distribution(stats)
      return {} if stats.total.zero?

      stats.value_counts.sort_by { |v, n| [-n, v] }.to_h.transform_values do |n|
        (n.to_f / stats.total).round(4)
      end
    end

    # subtype_distribution slices type_counts to a specific subset and
    # returns the fraction each subtype represents. Used for the :number
    # umbrella to expose the int-vs-float split.
    def subtype_distribution(stats, subtypes)
      return {} if stats.total.zero?

      subtypes.each_with_object({}) do |t, out|
        n = stats.type_counts[t] || 0
        out[t] = (n.to_f / stats.total).round(4) if n.positive?
      end
    end

    # file_kind_distribution buckets tracked values by file kind and
    # returns the fraction each kind represents over tracked observations.
    # `:unknown` covers values that classified as :file but whose extension
    # isn't in the kind allowlist (shouldn't normally happen since the
    # classifier already gates on the kind map). Sums to ≤ 1.0 since
    # value_counts caps at PositionStats::DEFAULT_MAX_VALUES.
    def file_kind_distribution(stats)
      return {} if stats.value_counts.empty?

      total = stats.value_counts.values.sum
      return {} if total.zero?

      kinds = Hash.new(0)
      stats.value_counts.each do |value, n|
        kind = SegmentClassifier.file_kind(value) || :unknown
        kinds[kind] += n
      end
      kinds.sort_by { |k, n| [-n, k.to_s] }.to_h.transform_values do |n|
        (n.to_f / total).round(4)
      end
    end

    # Most common type in stats.type_counts excluding `skip` — lex tie-break
    # so the choice is deterministic across runtimes.
    def dominant_excluding(stats, skip)
      best = nil
      best_count = -1
      stats.type_counts.each do |t, n|
        next if t == skip
        if n > best_count || (n == best_count && t.to_s < best.to_s)
          best = t
          best_count = n
        end
      end
      best
    end

    # JSON-friendly dump for persistence (distinct from #to_h which is a
    # display form). Examples are dumped as canonical strings and re-parsed
    # on load.
    def dump
      {
        "key"            => key,
        "host"           => host,
        "scheme"         => scheme,
        "shape"          => shape,
        "count"          => count,
        "examples"       => examples.map(&:canonical),
        "segment_counts" => @segment_counts.map { |h| h || {} },
        "param_stats"    => @param_stats.transform_values(&:dump),
      }
    end

    def self.from_dump(h, max_values: PositionStats::DEFAULT_MAX_VALUES)
      cluster = new(
        key: h["key"], host: h["host"], scheme: h["scheme"], shape: h["shape"],
        max_values: max_values,
      )
      cluster.instance_variable_set(:@count, h["count"])
      examples = h["examples"].map { |s| Parser.parse(s) }
      cluster.instance_variable_set(:@examples, examples)
      cluster.instance_variable_set(:@example_keys, examples.map(&:canonical).to_set)
      cluster.instance_variable_set(:@segment_counts, h["segment_counts"].map { |sub| Hash.new(0).merge(sub) })
      params = (h["param_stats"] || {}).transform_values { |sd| PositionStats.from_dump(sd) }
      cluster.instance_variable_set(:@param_stats, params)
      cluster
    end

    # Shared cluster-key derivation. Returns [key, host, scheme, shape] —
    # callers that already have a hinted shape can pass it in to skip the
    # recomputation; URN inputs ignore the override and always derive their
    # own shape from the NSS value. `host:` overrides iri.host — used by
    # Corpus when host_strategy collapses subdomains or ignores the host.
    def self.key_for(iri, classifier:, shape: nil, host: nil)
      if iri.urn?
        ns, value = (iri.nss || "").split(":", 2)
        derived = value ? urn_value_shape(ns, value, classifier) : nil
        key     = "urn:#{ns}:#{derived}"
        [key, nil, "urn", key]
      else
        shape ||= PathShape.new(classifier: classifier).for(iri.path_segments)
        effective_host = host.nil? ? iri.host : host
        key = "#{iri.scheme}://#{effective_host}#{shape}"
        [key, effective_host, iri.scheme, shape]
      end
    end

    def self.urn_value_shape(ns, value, classifier)
      entry = SegmentHints.derive([ns, value], classifier).last
      return entry[:value] unless entry[:variable]

      "{#{entry[:hint] || entry[:type]}}"
    end
  end
end

File: rust/iriq/src/cluster.rs
use crate::classifier::{
    file_kind, param_name_hint, FileKind, SegmentClassifier, SegmentType, DEFAULT_CLASSIFIER,
};
use crate::hints::SegmentHint;
use crate::identifier::Identifier;
use crate::position_stats::{PositionStats, DEFAULT_MAX_VALUES_PER_POSITION};
use std::collections::{HashMap, HashSet};
use std::sync::Arc;

pub const MAX_CLUSTER_EXAMPLES: usize = 10;
pub const DATE_CONFIDENCE_THRESHOLD: f64 = 0.8;
pub const NUMBER_CONFIDENCE_THRESHOLD: f64 = 0.8;
pub const NUMBER_SUBTYPE_THRESHOLD: f64 = 0.8;

// Param classification is a confidence ladder: constant → string → enum. A
// single-valued param is a constant (rendered as-is); one that varies but
// isn't a trustworthy enum is SegmentType::String (a generic placeholder); a
// bounded, well-supported value set is SegmentType::Enum.
//
// An enum is promoted when there are enough samples to trust the bound
// (ENUM_MIN_OBSERVATIONS), the *established* values — those seen at least
// ENUM_MIN_VALUE_COUNT times — number between ENUM_MIN_MEMBERS and
// ENUM_MAX_CARDINALITY, and cover nearly all observations (ENUM_MIN_COVERAGE).
// Rare one-off values are stragglers, not disqualifiers, so a single brand-new
// value can't knock an established enum down (the observe-before-normalize case).
pub const ENUM_MIN_OBSERVATIONS: usize = 20;
pub const ENUM_MAX_CARDINALITY: usize = 10;
pub const ENUM_MIN_VALUE_COUNT: usize = 3;
pub const ENUM_MIN_COVERAGE: f64 = 0.9;
// An enum is a bounded *set*: a single repeated value is a constant, not an
// enum, so it takes at least two established members to qualify.
pub const ENUM_MIN_MEMBERS: usize = 2;

// A literal-valued param that has taken on at least this many distinct values
// varies, so it's SegmentType::String rather than a fixed constant.
pub const STRING_MIN_DISTINCT: usize = 2;

// confidence = total / (total + K): a monotone curve that is 0.5 at K
// observations and asymptotes to 1.0. The type names our guess; this number
// says how much evidence backs it.
pub const CONFIDENCE_SMOOTHING: usize = 15;

pub const YEAR_RANGE_MIN: f64 = 1900.0;
pub const YEAR_RANGE_MAX: f64 = 2100.0;
pub const YEAR_MIN_OBSERVATIONS: usize = 5;
pub const YEAR_MIN_DISTINCT: usize = 2;
pub const YEAR_MAX_DISTINCT: usize = 150;

pub const HTTP_STATUS_RANGE_MIN: f64 = 100.0;
pub const HTTP_STATUS_RANGE_MAX: f64 = 599.0;
pub const HTTP_STATUS_MIN_OBSERVATIONS: usize = 5;
pub const HTTP_STATUS_MIN_DISTINCT: usize = 2;
pub const HTTP_STATUS_MAX_DISTINCT: usize = 30;

#[derive(Debug, Clone)]
pub struct SegmentPositionStat {
    pub position: usize,
    pub stable: bool,
    pub values: HashMap<String, usize>,
}

#[derive(Debug, Clone)]
pub struct Cluster {
    pub key: String,
    pub host: String,
    pub scheme: String,
    pub shape: String,
    pub examples: Vec<Arc<Identifier>>,
    pub count: usize,
    pub segment_counts: Vec<HashMap<String, usize>>,
    pub param_stats: HashMap<String, PositionStats>,
    pub max_values: usize,
    pub example_keys: HashSet<String>,
}

impl Cluster {
    pub fn new(
        key: String,
        host: String,
        scheme: String,
        shape: String,
        max_values: usize,
    ) -> Self {
        let cap = if max_values == 0 {
            DEFAULT_MAX_VALUES_PER_POSITION
        } else {
            max_values
        };
        Cluster {
            key,
            host,
            scheme,
            shape,
            examples: Vec::new(),
            count: 0,
            segment_counts: Vec::new(),
            param_stats: HashMap::new(),
            max_values: cap,
            example_keys: HashSet::new(),
        }
    }

    pub fn add(&mut self, iri: &Identifier) {
        self.add_with(iri, &DEFAULT_CLASSIFIER)
    }

    pub fn add_with(&mut self, iri: &Identifier, classifier: &SegmentClassifier) {
        self.count += 1;
        if self.examples.len() < MAX_CLUSTER_EXAMPLES {
            let canon = iri.canonical();
            if self.example_keys.insert(canon) {
                self.examples.push(Arc::new(iri.clone()));
            }
        }
        for (i, seg) in iri.path_segments.iter().enumerate() {
            while self.segment_counts.len() <= i {
                self.segment_counts.push(HashMap::new());
            }
            *self.segment_counts[i].entry(seg.clone()).or_insert(0) += 1;
        }
        for (name, v) in iri.query_params.iter() {
            let stats = self
                .param_stats
                .entry(name.to_string())
                .or_insert_with(|| PositionStats::new(self.max_values));
            stats.observe(v, classifier.classify(v));
        }
    }

    pub fn register_example_key(&mut self, canon: String) {
        self.example_keys.insert(canon);
    }

    pub fn segment_stats(&self) -> Vec<SegmentPositionStat> {
        self.segment_counts
            .iter()
            .enumerate()
            .map(|(i, counts)| SegmentPositionStat {
                position: i,
                stable: counts.len() == 1,
                values: counts.clone(),
            })
            .collect()
    }

    pub fn param_summary(&self) -> Vec<ParamSummary> {
        if self.param_stats.is_empty() {
            return Vec::new();
        }
        let mut rows: Vec<ParamSummary> = self
            .param_stats
            .iter()
            .map(|(name, stats)| {
                let presence = if self.count > 0 {
                    (stats.total as f64) / (self.count as f64)
                } else {
                    0.0
                };
                let ty = self.param_type(name);
                let mut row = ParamSummary {
                    name: name.clone(),
                    count: stats.total,
                    ty,
                    confidence: param_confidence(stats),
                    cardinality: stats.cardinality(),
                    presence,
                    values: Vec::new(),
                    numeric_count: 0,
                    min: 0.0,
                    max: 0.0,
                    avg: 0.0,
                    value_distribution: HashMap::new(),
                    subtype_distribution: HashMap::new(),
                    kind_distribution: HashMap::new(),
                };
                if row.ty == SegmentType::Enum {
                    row.values = enum_values(stats);
                }
                if row.ty == SegmentType::Boolean || row.ty == SegmentType::Enum {
                    row.value_distribution = value_distribution(stats);
                }
                if row.ty == SegmentType::Number {
                    row.subtype_distribution =
                        subtype_distribution(stats, &[SegmentType::Integer, SegmentType::Float]);
                }
                if row.ty == SegmentType::File {
                    row.kind_distribution = file_kind_distribution(stats);
                }
                if stats.numeric_count > 0 {
                    row.numeric_count = stats.numeric_count;
                    row.min = stats.numeric_min;
                    row.max = stats.numeric_max;
                    row.avg = stats.numeric_avg();
                }
                row
            })
            .collect();
        sort_param_summary(&mut rows);
        rows
    }

    pub fn param_type(&self, name: &str) -> SegmentType {
        let stats = match self.param_stats.get(name) {
            Some(s) if s.total > 0 => s,
            _ => return SegmentType::Literal,
        };
        let t = stats.dominant_type();

        if is_year_position(t, stats) {
            return SegmentType::Year;
        }
        if is_http_status_position(t, stats) {
            return SegmentType::HttpStatus;
        }

        if is_enum(stats) && t != SegmentType::Boolean {
            return SegmentType::Enum;
        }

        if t == SegmentType::Date {
            let date_frac = (*stats.type_counts.get(&SegmentType::Date).unwrap_or(&0) as f64)
                / (stats.total as f64);
            if date_frac >= DATE_CONFIDENCE_THRESHOLD {
                return t;
            }
            if let Some(alt) = dominant_excluding(stats, SegmentType::Date) {
                return alt;
            }
            return SegmentType::Literal;
        }

        if t == SegmentType::Integer || t == SegmentType::Float {
            let int_frac = (*stats.type_counts.get(&SegmentType::Integer).unwrap_or(&0) as f64)
                / (stats.total as f64);
            let float_frac = (*stats.type_counts.get(&SegmentType::Float).unwrap_or(&0) as f64)
                / (stats.total as f64);
            if int_frac < NUMBER_SUBTYPE_THRESHOLD
                && float_frac < NUMBER_SUBTYPE_THRESHOLD
                && (int_frac + float_frac) >= NUMBER_CONFIDENCE_THRESHOLD
            {
                return SegmentType::Number;
            }
        }

        if let Some(hint) = param_name_hint(name, t) {
            return hint;
        }

        // String rung — a literal-valued param that has taken on more than one
        // distinct value varies, so it's a placeholder, not a fixed constant.
        // Below the enum bar (checked above), so we claim only "free-form text".
        if t == SegmentType::Literal && stats.cardinality() >= STRING_MIN_DISTINCT {
            return SegmentType::String;
        }
        t
    }
}

#[derive(Debug, Clone)]
pub struct ParamSummary {
    pub name: String,
    pub count: usize,
    pub ty: SegmentType,
    pub confidence: f64,
    pub cardinality: usize,
    pub presence: f64,
    pub values: Vec<String>,
    pub numeric_count: usize,
    pub min: f64,
    pub max: f64,
    pub avg: f64,
    pub value_distribution: HashMap<String, f64>,
    pub subtype_distribution: HashMap<SegmentType, f64>,
    pub kind_distribution: HashMap<FileKind, f64>,
}

fn round_frac(f: f64) -> f64 {
    (f * 10000.0).round() / 10000.0
}

pub fn value_distribution(stats: &PositionStats) -> HashMap<String, f64> {
    if stats.total == 0 {
        return HashMap::new();
    }
    stats
        .value_counts
        .iter()
        .map(|(v, n)| (v.clone(), round_frac((*n as f64) / (stats.total as f64))))
        .collect()
}

pub fn subtype_distribution(
    stats: &PositionStats,
    subtypes: &[SegmentType],
) -> HashMap<SegmentType, f64> {
    if stats.total == 0 {
        return HashMap::new();
    }
    let mut out = HashMap::new();
    for &t in subtypes {
        let n = *stats.type_counts.get(&t).unwrap_or(&0);
        if n > 0 {
            out.insert(t, round_frac((n as f64) / (stats.total as f64)));
        }
    }
    out
}

pub fn file_kind_distribution(stats: &PositionStats) -> HashMap<FileKind, f64> {
    if stats.value_counts.is_empty() {
        return HashMap::new();
    }
    let total: usize = stats.value_counts.values().sum();
    if total == 0 {
        return HashMap::new();
    }
    let mut counts: HashMap<Option<FileKind>, usize> = HashMap::new();
    for (v, n) in &stats.value_counts {
        let k = file_kind(v);
        *counts.entry(k).or_insert(0) += *n;
    }
    let mut out = HashMap::new();
    for (k, n) in counts {
        // Unknown values are bucketed separately in Ruby under :unknown;
        // here we filter to known kinds. (Phase 1 omitted this nuance; phase 2 keeps it.)
        if let Some(kind) = k {
            out.insert(kind, round_frac((n as f64) / (total as f64)));
        }
    }
    out
}

// The enum's member values — the established ones (seen enough to be real),
// ordered by descending count with a lex tie-break. Stragglers are excluded so
// the advertised set is what the corpus is actually confident about.
pub fn enum_values(stats: &PositionStats) -> Vec<String> {
    let mut keys: Vec<String> = stats
        .value_counts
        .iter()
        .filter(|(_, &n)| n >= ENUM_MIN_VALUE_COUNT)
        .map(|(k, _)| k.clone())
        .collect();
    keys.sort_by(|a, b| {
        let na = stats.value_counts[a];
        let nb = stats.value_counts[b];
        nb.cmp(&na).then(a.cmp(b))
    });
    keys
}

// Built around the *established* members (values seen at least
// ENUM_MIN_VALUE_COUNT times) so a stray one-off value is a straggler, not a
// disqualifier.
pub fn is_enum(stats: &PositionStats) -> bool {
    if stats.total < ENUM_MIN_OBSERVATIONS {
        return false;
    }
    let mut established = 0usize; // established members
    let mut covered = 0usize; // observations they account for
    for &n in stats.value_counts.values() {
        if n >= ENUM_MIN_VALUE_COUNT {
            established += 1;
            covered += n;
        }
    }
    if !(ENUM_MIN_MEMBERS..=ENUM_MAX_CARDINALITY).contains(&established) {
        return false;
    }
    (covered as f64) / (stats.total as f64) >= ENUM_MIN_COVERAGE
}

// How much evidence backs the assigned type: monotone in observation count,
// 0.5 at CONFIDENCE_SMOOTHING, asymptoting to 1.0. Rounded to two decimals to
// match the Ruby/Go output.
pub fn param_confidence(stats: &PositionStats) -> f64 {
    if stats.total == 0 {
        return 0.0;
    }
    let c = (stats.total as f64) / ((stats.total + CONFIDENCE_SMOOTHING) as f64);
    (c * 100.0).round() / 100.0
}

pub fn is_year_position(t: SegmentType, stats: &PositionStats) -> bool {
    if t != SegmentType::Integer || stats.numeric_count == 0 {
        return false;
    }
    let card = stats.cardinality();
    if !(YEAR_MIN_DISTINCT..=YEAR_MAX_DISTINCT).contains(&card) {
        return false;
    }
    if stats.total < YEAR_MIN_OBSERVATIONS {
        return false;
    }
    stats.numeric_min >= YEAR_RANGE_MIN
        && stats.numeric_min <= YEAR_RANGE_MAX
        && stats.numeric_max >= YEAR_RANGE_MIN
        && stats.numeric_max <= YEAR_RANGE_MAX
}

pub fn is_http_status_position(t: SegmentType, stats: &PositionStats) -> bool {
    if t != SegmentType::Integer || stats.numeric_count == 0 {
        return false;
    }
    let card = stats.cardinality();
    if !(HTTP_STATUS_MIN_DISTINCT..=HTTP_STATUS_MAX_DISTINCT).contains(&card) {
        return false;
    }
    if stats.total < HTTP_STATUS_MIN_OBSERVATIONS {
        return false;
    }
    stats.numeric_min >= HTTP_STATUS_RANGE_MIN
        && stats.numeric_min <= HTTP_STATUS_RANGE_MAX
        && stats.numeric_max >= HTTP_STATUS_RANGE_MIN
        && stats.numeric_max <= HTTP_STATUS_RANGE_MAX
}

pub fn dominant_excluding(stats: &PositionStats, skip: SegmentType) -> Option<SegmentType> {
    let mut best: Option<(SegmentType, usize)> = None;
    for (&t, &n) in &stats.type_counts {
        if t == skip {
            continue;
        }
        best = match best {
            None => Some((t, n)),
            Some((bt, bn)) => {
                if n > bn || (n == bn && t.as_str() < bt.as_str()) {
                    Some((t, n))
                } else {
                    Some((bt, bn))
                }
            }
        };
    }
    best.map(|(t, _)| t)
}

fn sort_param_summary(rows: &mut [ParamSummary]) {
    rows.sort_by(|a, b| b.count.cmp(&a.count).then(a.name.cmp(&b.name)));
}

// Conveniences used by Cluster / Corpus when deriving keys.
pub fn placeholder_for(e: &SegmentHint) -> String {
    if !e.variable {
        return e.value.clone();
    }
    if !e.hint.is_empty() {
        return format!("{{{}}}", e.hint);
    }
    format!("{{{}}}", e.ty.as_str())
}
